README.md
[![Build Status](https://travis-ci.org/mapbox/tilelive-memcached.png?branch=master)](https://travis-ci.org/mapbox/tilelive-memcached)

tilelive-memcached
------------------
node-tilejson wrapping source for tilelive.

    var options = {
        mode: 'readthrough', // optional, cache mode either 'readthrough' or 'race'
        client: client, // optional, instantiated memcached client
        expires: 600    // optional, object expiration time in seconds
        ttl: 300        // optional, relay mode only, numbe of seconds before an object should be re-checked.
    };
    var TileJSON = require('tilelive-memcached')(options, require('tilejson'));

    new TileJSON( ... )

### Cache modes

Two modes for caching are available.

- **readthrough** hits memcached first and only calls a `get` on the original source if a cache miss occurs.
- **race** always hits both memcached and the original source concurrently. The IO operation that completes fastest will handle the `get` call. After both operations are complete the cache may be updated if the original source's contents have changed.


CHANGELOG.md
## 1.0.1

- Adding 'relay' cache mode which serves available but stale data while refreshing in the background.

## 1.0.0

- Switched to binary client.
- Removed node 0.8.x support.

## 0.x

- Initial series. Supported "racing" and "readthrough" cache modes. Used ASCII client.


test/testsource.js
module.exports = Testsource;

var now = new Date;
var tiles = {
    a: require('fs').readFileSync(__dirname + '/a.png'),
    b: require('fs').readFileSync(__dirname + '/b.png'),
};

var grids = {
    a: { grid:'', keys: ['', '1' ], data:{'1': {'name':'foo'}} },
    b: { grid:'', keys: ['', '1' ], data:{'1': {'name':'bar'}} },
};

Testsource.now = new Date;
Testsource.tiles = tiles;
Testsource.grids = grids;

// Define a mock test source.
function Testsource(uri, callback) {
    this._uri = uri;
    this.delay = uri.delay || 0;
    this.hostname = uri.hostname || 'test';
    this.stat = {
        'get': 0
    };
    callback(null, this);
}
Testsource.prototype.get = function(url, callback) {
    var stat = this.stat;

    if (this.delay) {
        setTimeout(function() { get(url, callback); }, this.delay);
    } else {
        get();
    }

    function get() {
        stat.get++;
        switch (url) {
        case 'http://test/0/0/0.png':
            return callback(null, tiles.a, {
                'content-type': 'image/png',
                'content-length': 11541,
                'last-modified': now.toUTCString()
            });
        case 'http://test/1/0/0.png':
            return callback(null, tiles.b, {
                'content-type': 'image/png',
                'content-length': 6199,
                'last-modified': now.toUTCString()
            });
        case 'http://test/2/0/0.png':
            return callback(new Error('Unexpected error'));
        case 'http://test/0/0/0.grid.json':
            return callback(null, JSON.stringify(grids.a), {
                'content-type': 'application/json',
                'last-modified': now.toUTCString()
            });
        case 'http://test/1/0/0.grid.json':
            return callback(null, JSON.stringify(grids.b), {
                'content-type': 'application/json',
                'last-modified': now.toUTCString()
            });
        case 'http://long/0/0/0.png':
            return callback(null, tiles.a, {
                'content-type': 'image/png',
                'content-length': 11541,
                'last-modified': now.toUTCString()
            });
        case 'http://long/1/0/0.png':
            return callback(null, tiles.b, {
                'content-type': 'image/png',
                'content-length': 6199,
                'last-modified': now.toUTCString()
            });
        case 'http://long/0/0/0.grid.json':
            return callback(null, JSON.stringify(grids.a), {
                'content-type': 'application/json',
                'last-modified': now.toUTCString()
            });
        case 'http://long/1/0/0.grid.json':
            return callback(null, JSON.stringify(grids.b), {
                'content-type': 'application/json',
                'last-modified': now.toUTCString()
            });
        case 'http://stale/0/0/0.png':
            return callback(null, tiles.a, {
                'content-type': 'image/png',
                'content-length': 11541,
                'last-modified': now.toUTCString()
            });
        case 'http://stale/1/0/0.png':
            return callback(null, tiles.b, {
                'content-type': 'image/png',
                'content-length': 6199,
                'last-modified': now.toUTCString()
            });
        case 'http://stale/0/0/0.grid.json':
            return callback(null, JSON.stringify(grids.a), {
                'content-type': 'application/json',
                'last-modified': now.toUTCString()
            });
        case 'http://stale/1/0/0.grid.json':
            return callback(null, JSON.stringify(grids.b), {
                'content-type': 'application/json',
                'last-modified': now.toUTCString()
            });
        default:
            var err = new Error('Not found');
            err.statusCode = 404;
            return callback(err);
        }
    }
};
Testsource.prototype.getTile = function(z, x, y, callback) {
    this.get('http://' + this.hostname + '/' + [z,x,y].join('/') + '.png', function(err, buffer, headers) {
        if (err) {
            return callback(err);
        }
        return callback(null, buffer, headers);
    });
};
Testsource.prototype.getGrid = function(z, x, y, callback) {
    this.get('http://' + this.hostname + '/' + [z,x,y].join('/') + '.grid.json', function(err, buffer, headers) {
        if (err) {
            return callback(err);
        }
        return callback(null, JSON.parse(buffer), headers);
    });
};



index.js
var urlParse = require('url').parse;
var util = require('util');
var memjs = require('memjs');

module.exports = function(options, Source) {
    if (!Source) throw new Error('No source provided');
    if (!Source.prototype.get) throw new Error('No get method found on source');

    function Caching() { return Source.apply(this, arguments); }

    // Inheritance.
    util.inherits(Caching, Source);

    // References for testing, convenience, post-call overriding.
    Caching.memcached = options;

    Caching.prototype.get = module.exports.cachingGet('TL', options, Source.prototype.get);

    return Caching;
};

module.exports.cachingGet = function(namespace, options, get) {
    if (!get) throw new Error('No get function provided');
    if (!namespace) throw new Error('No namespace provided');

    options = options || {};
    options.client = ('client' in options) ? options.client : memjs.Client.create(null, {'keepAlive': true});
    options.expires = ('expires' in options) ? options.expires : 300;
    options.mode = ('mode' in options) ? options.mode : 'readthrough';
    options.logger = ('logger' in options) ? options.logger: console;

    if (!options.client) throw new Error('No memcached client');
    if (!options.expires) throw new Error('No expires option set');

    var caching;
    if (options.mode === 'readthrough') {
        caching = readthrough;
    } else if (options.mode === 'race') {
        caching = race;
    } else {
        throw new Error('Invalid value for options.mode ' + options.mode);
    }

    function race(url, callback) {
        var key = namespace + '-' + url;
        var source = this;
        var client = options.client;
        var expires;
        if (typeof options.expires === 'number') {
            expires = options.expires;
        } else {
            expires = options.expires[urlParse(url).hostname] || options.expires.default || 300;
        }

        var sent = false;
        var cached = null;
        var current = null;

        // GET upstream.
        get.call(source, url, function(err, buffer, headers) {
            current = encode(err, buffer, headers);
            if (cached && current) finalize();
            if (sent) return;
            sent = true;
            callback(err, buffer, headers);
        });

        // GET memcached.
        client.get(key, function(err, encoded) {
            // If error on memcached, do not flip first flag.
            // Finalize will never occur (no cache set).
            if (err) {
                err.key = key;
                return options.logger.log(err);
            }

            cached = encoded || '500';
            if (cached && current) finalize();
            if (sent || !encoded) return;
            var data;
            try {
                data = decode(cached);
            } catch(e) {
                e.key = key;
                options.logger.log(e);
                cached = '500';
            }
            if (data) {
                sent = true;
                callback(data.err, data.buffer, data.headers);
            }
        });

        function finalize() {
            if (cached === current) return;
            client.set(key, current, function(err) {
                if (err) {
                    err.key = key;
                    options.logger.log(err);
                }
            }, expires);
        }
    }

    function readthrough(url, callback) {
        var key = namespace + '-' + url;
        var source = this;
        var client = options.client;
        var expires;
        if (typeof options.expires === 'number') {
            expires = options.expires;
        } else {
            expires = options.expires[urlParse(url).hostname] || options.expires.default || 300;
        }
        client.get(key, function(err, encoded) {
            // If error on memcached get, pass through to original source
            // without attempting a set after retrieval.
            if (err) {
                err.key = key;
                options.logger.log(err);
                return get(url, callback);
            }

            // Cache hit.
            var data;
            if (encoded) try {
                data = decode(encoded);
            } catch(e) {
                e.key = key;
                options.logger.log(e);
            }
            if (data) return callback(data.err, data.buffer, data.headers);

            // Cache miss, error, or otherwise no data
            get.call(source, url, function(err, buffer, headers) {
                if (err && !errcode(err)) return callback(err);
                callback(err, buffer, headers);

                // Callback does not need to wait for memcached set to occur.
                client.set(key, encode(err, buffer, headers), function(err) {
                    if (!err) return;
                    err.key = key;
                    options.logger.log(err);
                }, expires);
            });
        });
    }

    return caching;
};

module.exports.memjs = memjs;
module.exports.encode = encode;
module.exports.decode = decode;

function errcode(err) {
    if (!err) return;
    if (err.statusCode === 404) return 404;
    if (err.statusCode === 403) return 403;
    return;
}

function encode(err, buffer, headers) {
    if (errcode(err)) return errcode(err).toString();

    // Unhandled error.
    if (err) return null;

    headers = headers || {};

    // Turn objects into JSON string buffers.
    if (buffer && typeof buffer === 'object' && !(buffer instanceof Buffer)) {
        headers['x-memcached-json'] = true;
        buffer = new Buffer(JSON.stringify(buffer));
    // Turn strings into buffers.
    } else if (buffer && !(buffer instanceof Buffer)) {
        buffer = new Buffer(buffer);
    }

    headers = new Buffer(JSON.stringify(headers), 'utf8');

    if (headers.length > 1024) {
        throw new Error('Invalid cache value - headers exceed 1024 bytes: ' + JSON.stringify(headers));
    }

    var padding = new Buffer(1024 - headers.length);
    padding.fill(' ');
    var len = headers.length + padding.length + buffer.length;
    return Buffer.concat([headers, padding, buffer], len);
}

function decode(encoded) {
    if (encoded.length == 3) {
        encoded = encoded.toString();
        if (encoded === '404' || encoded === '403') {
            var err = new Error(encoded === '404' ? 'Not found' : 'forbidden');
            err.statusCode = parseInt(encoded, 10);
            err.memcached = true;
            return { err: err };
        }
    }

    // First 1024 bytes reserved for header + padding.
    var offset = 1024;
    var data = {};
    data.headers = encoded.slice(0, offset).toString().trim();

    try {
        data.headers = JSON.parse(data.headers);
    } catch(e) {
        throw new Error('Invalid cache value');
    }

    data.headers['x-memcached'] = 'hit';
    data.buffer = encoded.slice(offset);

    // Return JSON-encoded objects to true form.
    if (data.headers['x-memcached-json']) data.buffer = JSON.parse(data.buffer);

    if (data.headers['content-length'] && data.headers['content-length'] != data.buffer.length)
        throw new Error('Content length does not match');
    return data;
}
